File: leet_BFS_2583_kth_largest.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function kthLargestLevelSum(root: TreeNode | null, k: number): number {
    let D:TreeNode[] = [root]
    let res:number[] = []
    while (D.length) {
        let level:number = 0
        let len:number = D.length
        while (len--) {
            let node:TreeNode = D.shift()
            level += node.val
            if (node.left)
                D.push(node.left)
            if (node.right)
                D.push(node.right)
        }
        res.push(level)
    }
    let n:number = res.length
    if (n < k) {
        return -1
    }

    // res.sort((a, b) => {return b - a})
    // return res[k - 1]

    res.sort((a, b) => {return a - b})
    return res[n - k]

};

File: leet_BFS_2583_kth_largest.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func kthLargestLevelSum(root *TreeNode, k int) int64 {
    solutions := []func(*TreeNode, int)int64{
        BFS,
    }
    return solutions[0]( root, k )
}

func BFS(root *TreeNode, k int) int64 {
	D := []*TreeNode{root}
	res := []int64{}
	for len(D) > 0 {
		var level int64 = 0
	    N := len(D)
		for N > 0 {
			node := D[0]
			D = D[1:]
			if node.Left != nil { D = append(D, node.Left) }
			if node.Right != nil { D = append(D, node.Right) }
            level += int64(node.Val)
            N--
		}
		res = append(res, level)
	}

	N := len(res)
	if N < k {
		return -1
	}
	sort.Slice(res, func(i, j int) bool {
		return res[i] > res[j]
	})
	return res[k - 1]

    // sort.Ints(res) // bc. it's a int64 slice so we can't use Ints
    // return res[N - k]

}
